### ISSUE TITLE
Multiple possible memory safety issues

### ISSUE THREAD

**Issue by fa1lr4in** [ImageMagick/ImageMagick#4452]
### ImageMagick version

7.1.0-14

### Operating system

Linux

### Operating system, version and so on

OS: Ubuntu 18.04.6 LTS  Version: ImageMagick 7.1.0-14 Q16-HDRI x86_64 2021-10-28 https://imagemagick.org Copyright: (C) 1999-2021 ImageMagick Studio LLC License: https://imagemagick.org/script/license.php Features: Cipher DPC HDRI OpenMP(5.0)  Delegates (built-in): bzlib fontconfig freetype jbig jng jpeg lcms lzma pangocairo png tiff webp x xml zlib Compiler: gcc (4.2)

### Description

When I use the fuzzing tool, I have detected some possible memory security issues through ASAN. I am not sure if they are security vulnerabilities. They have four different function call stacks.

### Steps to Reproduce

build it
CFLAGS="-g -fsanitize=address -fno-omit-frame-pointer" CXXFLAGS="-g -fsanitize=address -fno-omit-frame-pointer" LDFLAGS="-g -fsanitize=address -fno-omit-frame-pointer"
`./configure --disable-shared && make && make install
`
run it
`
./magick convert -adjoin -alpha copy -antialias -append -auto-gamma -auto-level  -auto-orient poc  /dev/null
`
The function call stack is as follows
poc1
==43761==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 6400 byte(s) in 2 object(s) allocated from:
    #0 0x5d1c77 in posix_memalign (/root/fuzz/target/imagemagick/ImageMagick/install/bin/magick+0x5d1c77)
    #1 0x8a56be in AcquireAlignedMemory_POSIX /root/fuzz/target/imagemagick/ImageMagick/MagickCore/memory.c:299:7
    #2 0x8a56be in AcquireAlignedMemory /root/fuzz/target/imagemagick/ImageMagick/MagickCore/memory.c:377:10
    #3 0x6971ad in AcquireCacheNexusPixels /root/fuzz/target/imagemagick/ImageMagick/MagickCore/cache.c:4977:37
    #4 0x6971ad in SetPixelCacheNexusPixels /root/fuzz/target/imagemagick/ImageMagick/MagickCore/cache.c:5097:12
    #5 0x690fe0 in QueueAuthenticPixelCacheNexus /root/fuzz/target/imagemagick/ImageMagick/MagickCore/cache.c:4186:10
    #6 0x714235 in QueueCacheViewAuthenticPixels /root/fuzz/target/imagemagick/ImageMagick/MagickCore/cache-view.c:990:10
    #7 0xb7431d in .omp_outlined._debug__.13 /root/fuzz/target/imagemagick/ImageMagick/MagickCore/transform.c:2311:7
    #8 0xb7431d in .omp_outlined..15.20308 /root/fuzz/target/imagemagick/ImageMagick/MagickCore/transform.c:2291:3
    #9 0x7f3419775892 in __kmp_invoke_microtask (/usr/lib/x86_64-linux-gnu/libomp.so.5+0xba892)
    #10 0x7f34196fedc2  (/usr/lib/x86_64-linux-gnu/libomp.so.5+0x43dc2)

SUMMARY: AddressSanitizer: 6400 byte(s) leaked in 2 allocation(s).

poc2
==43778==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 5740 byte(s) in 1 object(s) allocated from:
    #0 0x5d11bd in malloc (/root/fuzz/target/imagemagick/ImageMagick/install/bin/magick+0x5d11bd)
    #1 0x147116e in ReadEPTImage /root/fuzz/target/imagemagick/ImageMagick/coders/ept.c:213:41
    #2 0x669878 in ReadImage /root/fuzz/target/imagemagick/ImageMagick/MagickCore/constitute.c:563:15
    #3 0x66e513 in ReadImages /root/fuzz/target/imagemagick/ImageMagick/MagickCore/constitute.c:955:9
    #4 0x1615cc5 in ConvertImageCommand /root/fuzz/target/imagemagick/ImageMagick/MagickWand/convert.c:611:18
    #5 0x17d56d0 in MagickCommandGenesis /root/fuzz/target/imagemagick/ImageMagick/MagickWand/mogrify.c:188:14
    #6 0x605d84 in MagickMain /root/fuzz/target/imagemagick/ImageMagick/utilities/magick.c:150:10
    #7 0x605d84 in main /root/fuzz/target/imagemagick/ImageMagick/utilities/magick.c:182:10
    #8 0x7fe23fe85bf6 in __libc_start_main /build/glibc-S9d2JN/glibc-2.27/csu/../csu/libc-start.c:310

Direct leak of 362 byte(s) in 1 object(s) allocated from:
    #0 0x5d11bd in malloc (/root/fuzz/target/imagemagick/ImageMagick/install/bin/magick+0x5d11bd)
    #1 0x147119c in ReadEPTImage /root/fuzz/target/imagemagick/ImageMagick/coders/ept.c:219:35
    #2 0x669878 in ReadImage /root/fuzz/target/imagemagick/ImageMagick/MagickCore/constitute.c:563:15
    #3 0x66e513 in ReadImages /root/fuzz/target/imagemagick/ImageMagick/MagickCore/constitute.c:955:9
    #4 0x1615cc5 in ConvertImageCommand /root/fuzz/target/imagemagick/ImageMagick/MagickWand/convert.c:611:18
    #5 0x17d56d0 in MagickCommandGenesis /root/fuzz/target/imagemagick/ImageMagick/MagickWand/mogrify.c:188:14
    #6 0x605d84 in MagickMain /root/fuzz/target/imagemagick/ImageMagick/utilities/magick.c:150:10
    #7 0x605d84 in main /root/fuzz/target/imagemagick/ImageMagick/utilities/magick.c:182:10
    #8 0x7fe23fe85bf6 in __libc_start_main /build/glibc-S9d2JN/glibc-2.27/csu/../csu/libc-start.c:310

SUMMARY: AddressSanitizer: 6102 byte(s) leaked in 2 allocation(s).

poc3
==36065==ERROR: AddressSanitizer: UNKNOWN SIGNAL on unknown address 0x000000000000 (pc 0x000000696d32 bp 0x7f9a7bb3d510 sp 0x7f9a7bb3d460 T21)
[1m[0m    #0 0x696d32 in SetPixelCacheNexusPixels /root/fuzz/target/imagemagick/ImageMagick/MagickCore/cache.c:5040:3
    #1 0x6a6eb7 in GetVirtualPixelCacheNexus /root/fuzz/target/imagemagick/ImageMagick/MagickCore/cache.c:2773:10
    #2 0x6a89ac in GetVirtualPixelCacheNexus /root/fuzz/target/imagemagick/ImageMagick/MagickCore/cache.c:3076:9
    #3 0x6a89ac in GetVirtualPixelCacheNexus /root/fuzz/target/imagemagick/ImageMagick/MagickCore/cache.c:3076:9
    #4 0x712cea in GetCacheViewVirtualPixels /root/fuzz/target/imagemagick/ImageMagick/MagickCore/cache-view.c:664:10
    #5 0x7b5ad8 in .omp_outlined._debug__.8 /root/fuzz/target/imagemagick/ImageMagick/MagickCore/composite.c:722:11
    #6 0x7b5ad8 in .omp_outlined..9 /root/fuzz/target/imagemagick/ImageMagick/MagickCore/composite.c:703:7
    #7 0x7f9a92b07892 in __kmp_invoke_microtask (/usr/lib/x86_64-linux-gnu/libomp.so.5+0xba892)
    #8 0x7f9a92a90dc2  (/usr/lib/x86_64-linux-gnu/libomp.so.5+0x43dc2)
    #9 0x7f9a92a8f749  (/usr/lib/x86_64-linux-gnu/libomp.so.5+0x42749)
    #10 0x7f9a92aea629  (/usr/lib/x86_64-linux-gnu/libomp.so.5+0x9d629)
    #11 0x7f9a8ee886da in start_thread /build/glibc-S9d2JN/glibc-2.27/nptl/pthread_create.c:463
    #12 0x7f9a8d86571e in clone /build/glibc-S9d2JN/glibc-2.27/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:95

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: UNKNOWN SIGNAL /root/fuzz/target/imagemagick/ImageMagick/MagickCore/cache.c:5040:3 in SetPixelCacheNexusPixels
Thread T21 created by T0 here:
    #0 0x5bb8bc in pthread_create (/root/fuzz/target/imagemagick/ImageMagick/install/bin/magick+0x5bb8bc)
    #1 0x7f9a92ae9cf3  (/usr/lib/x86_64-linux-gnu/libomp.so.5+0x9ccf3)

==36065==ABORTING

poc4
==43937==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6250000279ac at pc 0x0000009d2366 bp 0x7feab4161990 sp 0x7feab4161988
READ of size 4 at 0x6250000279ac thread T3
    #0 0x9d2365 in GetPixelAlpha /root/fuzz/target/imagemagick/ImageMagick/./MagickCore/pixel-accessor.h:60:10
    #1 0x9d2365 in .omp_outlined._debug__.64 /root/fuzz/target/imagemagick/ImageMagick/MagickCore/resize.c:3685:53
    #2 0x9d2365 in .omp_outlined..65.15622 /root/fuzz/target/imagemagick/ImageMagick/MagickCore/resize.c:3555:3
    #3 0x7feac214b892 in __kmp_invoke_microtask (/usr/lib/x86_64-linux-gnu/libomp.so.5+0xba892)
    #4 0x7feac20d4dc2  (/usr/lib/x86_64-linux-gnu/libomp.so.5+0x43dc2)
    #5 0x7feac20d3749  (/usr/lib/x86_64-linux-gnu/libomp.so.5+0x42749)
    #6 0x7feac212e629  (/usr/lib/x86_64-linux-gnu/libomp.so.5+0x9d629)
    #7 0x7feabe4cc6da in start_thread /build/glibc-S9d2JN/glibc-2.27/nptl/pthread_create.c:463
    #8 0x7feabcea971e in clone /build/glibc-S9d2JN/glibc-2.27/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:95

0x6250000279ac is located 172 bytes to the right of 8192-byte region [0x625000025900,0x625000027900)
allocated by thread T0 here:
    #0 0x5d1c77 in posix_memalign (/root/fuzz/target/imagemagick/ImageMagick/install/bin/magick+0x5d1c77)
    #1 0x8a56be in AcquireAlignedMemory_POSIX /root/fuzz/target/imagemagick/ImageMagick/MagickCore/memory.c:299:7
    #2 0x8a56be in AcquireAlignedMemory /root/fuzz/target/imagemagick/ImageMagick/MagickCore/memory.c:377:10
    #3 0x69a582 in OpenPixelCache /root/fuzz/target/imagemagick/ImageMagick/MagickCore/cache.c:3746:46
    #4 0x6960d9 in GetImagePixelCache /root/fuzz/target/imagemagick/ImageMagick/MagickCore/cache.c:1775:18
    #5 0x616b12 in SyncImagePixelCache /root/fuzz/target/imagemagick/ImageMagick/MagickCore/cache.c:5516:28
    #6 0x616b12 in SetImageExtent /root/fuzz/target/imagemagick/ImageMagick/MagickCore/image.c:2674:10
    #7 0x613468 in CloneImage /root/fuzz/target/imagemagick/ImageMagick/MagickCore/image.c:906:7
    #8 0x9c9912 in ResizeImage /root/fuzz/target/imagemagick/ImageMagick/MagickCore/resize.c:3790:18
    #9 0xf34329 in WriteMipmaps /root/fuzz/target/imagemagick/ImageMagick/coders/dds.c:3937:22
    #10 0xf34329 in WriteDDSImage /root/fuzz/target/imagemagick/ImageMagick/coders/dds.c:4192:25
    #11 0x675487 in WriteImage /root/fuzz/target/imagemagick/ImageMagick/MagickCore/constitute.c:1217:16
    #12 0x6786ee in WriteImages /root/fuzz/target/imagemagick/ImageMagick/MagickCore/constitute.c:1438:13
    #13 0x161ec15 in ConvertImageCommand /root/fuzz/target/imagemagick/ImageMagick/MagickWand/convert.c:3327:11
    #14 0x17d56d0 in MagickCommandGenesis /root/fuzz/target/imagemagick/ImageMagick/MagickWand/mogrify.c:188:14
    #15 0x605d84 in MagickMain /root/fuzz/target/imagemagick/ImageMagick/utilities/magick.c:150:10
    #16 0x605d84 in main /root/fuzz/target/imagemagick/ImageMagick/utilities/magick.c:182:10
    #17 0x7feabcda9bf6 in __libc_start_main /build/glibc-S9d2JN/glibc-2.27/csu/../csu/libc-start.c:310

Thread T3 created by T0 here:
    #0 0x5bb8bc in pthread_create (/root/fuzz/target/imagemagick/ImageMagick/install/bin/magick+0x5bb8bc)
    #1 0x7feac212dcf3  (/usr/lib/x86_64-linux-gnu/libomp.so.5+0x9ccf3)

SUMMARY: AddressSanitizer: heap-buffer-overflow /root/fuzz/target/imagemagick/ImageMagick/./MagickCore/pixel-accessor.h:60:10 in GetPixelAlpha
Shadow bytes around the buggy address:
  0x0c4a7fffcee0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c4a7fffcef0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c4a7fffcf00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c4a7fffcf10: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c4a7fffcf20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c4a7fffcf30: fa fa fa fa fa[fa]fa fa fa fa fa fa fa fa fa fa
  0x0c4a7fffcf40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4a7fffcf50: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4a7fffcf60: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4a7fffcf70: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4a7fffcf80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==43937==ABORTING

Thanks!

### Images

[poc1-4.zip](https://github.com/ImageMagick/ImageMagick/files/7495071/poc1-4.zip)

**Comment by urban-warrior** [comment]
Unfortunately, we cannot reproduce the issue for poc{1,3,4}.  We have gotten several similar reports,  all on Ubuntu that point to a compiler chain or libgomp issue.  We used your exact build commands on Fedora 35 and CentOS 8.  We also tried `valgrind` and it did not return any memory leaks, instead it returns `definitely lost: 0 bytes in 0 blocks`.

There is a leak in poc2 for the EPT coder.  This is not considered a security vulnerability but none-the-less we will add a patch to resolve the issue by sometime tomorrow.  Thanks for alerting us to this leak.
